governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,220 @@
+ * become unreferenced
+ * 
+ * @author tcellucci
+ *
+ */
+public class PreDestroyMonitor implements AutoCloseable {
+    
+     
+    private static class ScopeCleanupMarker {
+    }
+    
+    private final Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    
+    private final Map<Integer, ScopeCleanupAction> scopedCleanupActions = new LinkedHashMap<>();
+    private final Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private final ReferenceQueue<ScopeCleanupMarker> markerReferenceQueue = new ReferenceQueue<>();
+    private final ExecutorService reqQueueExecutor = Executors.newSingleThreadExecutor();
+    
+    private final Provider<ScopeCleanupMarker> markerProvider = new Provider<ScopeCleanupMarker>() {
+        @Override
+        public ScopeCleanupMarker get() {
+            return new ScopeCleanupMarker();
+        }
+    };
+    
+    public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
+        this.scopeBindings = new HashMap<>(scopeBindings);
+        reqQueueExecutor.submit(new Runnable() {
+            public void run() {
+                while (true) {
+                    try {
+                        Reference<? extends ScopeCleanupMarker> ref = markerReferenceQueue.remove(500);
+                        if (ref != null && ref instanceof ScopeCleanupAction) { 
+                            Integer markerKey = ((ScopeCleanupAction)ref).getId();
+                            ScopeCleanupAction cleanupAction;
+                            synchronized(scopedCleanupActions) {
+                                cleanupAction = scopedCleanupActions.remove(markerKey);
+                            }
+                            if (cleanupAction != null) {
+                                cleanupAction.call();
+                            }
+                        }
+                    } 
+                    catch (InterruptedException e) {
+                    }
+                }
+            }
+        });
+    }
+    
+    public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
+        return binding.acceptScopingVisitor(new ManagedInstanceScopingVisitor(destroyableInstance, binding.getSource(), action));
+    }
+    
+    public void addScopeBindings(Map<Class<? extends Annotation>, Scope> bindings) {
+        scopeBindings.putAll(bindings);
+    }
+
+    /*
+     * compatibility-mode - scope is assumed to be eager singleton
+     */
+    public <T> boolean register(T destroyableInstance, Object context, Iterable<LifecycleAction> action) {
+        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
+    }
+    
+    /**
+     * final cleanup of managed instances if any
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        if (!reqQueueExecutor.isShutdown()) reqQueueExecutor.shutdown();
```

Comment:
To improve readability please put the call to reqQueueExecutor.shutdown(); on the next line.